src/v/cluster/partition_balancer_planner.cc

```diff
@@ -163,6 +163,40 @@ class partition_balancer_planner::request_context {
     ss::abort_source& _as;
 };
 
+std::pair<uint64_t, uint64_t> partition_balancer_planner::get_node_bytes_info(
+  const node::local_state& node_state) {
+    const auto& state = node_state.log_data_size;
+    if (likely(state)) {
+        // It is okay to account reclaimable space as free space even through
+        // the space is not readily available. During the partition move if the
+        // disk is filled up, trimmer cleans up reclaimable space to make room
+        // for the move.
+        auto target_size = state.value().data_target_size;
+        auto current_size = state.value().data_current_size;
+        auto reclaimable_size = state.value().data_reclaimable_size;
+
+        auto actual_free = 0UL;
+        if (likely(target_size > current_size)) {
+            // to avoid any underflow with substraction with misreporting
+            actual_free = target_size - current_size;
+        }
+        // Clamp the free space to be conservative and account for any error
+        // from space monitoring logic.
+        actual_free = std::min(actual_free, node_state.data_disk.free);
+        auto total_free = actual_free + reclaimable_size;
+        // Clamp to total available target size.
```

Comment:
Edge case, but it seems like if target_size is low and we do have reclaimable bytes, we may return non-zero when instead we should return zero. E.g.

target_size = 5
current_size = 100
reclaimable_size = 50

we end up returning {5, 5}

Maybe something like

```
total_free = 0
if (target_size > current_size) {
  total_free = target_size - current_size + reclaimable_size;
} else {
  auto overage = current_size - target_size;
  if (reclaimable_size > overage) {
    total_free = reclaimable_size - overage;
  }
}
total_free = min(target_size, total_free)
```

Reply:
so i think we should cap at the target size. if the reclaimable is larger than target it it's a reasonable intermediate (but bad state) but the we want the balancer to only consider that the node itself has a total of 5 gb available, so it makes sense to cap that free space at 5 indepdnent of the reclaimable.

Reply:
oh wait. maybe i have it backwards. if current reported size - reclaimable size is larger than the target then it's like saying that "we've overshot the target, and if we trimmed everything, we'd still be over the target, so there is no free space"

Reply:
yeh so this seems right andrew. good spot